File: src/cadenza/java/cadenza/nbe/NeutralException.java
package cadenza.nbe;

import cadenza.types.Type;
import cadenza.values.NeutralValue;
import com.oracle.truffle.api.nodes.SlowPathException;

// this is used to suck the program through a straw using normalization-by-evaluation
// to produce a beta-eta-long normal form versions of the resulting program that remains

public class NeutralException extends SlowPathException {
  private static final long serialVersionUID = 5587798688299594259L;
  public final Type type;
  public final Neutral term;
  public NeutralValue get() { return new NeutralValue(type, term); }
  public NeutralException(Type type, Neutral term) { this.type = type; this.term = term; }
  public NeutralException apply(final Object... rands) {
    int len = rands.length;
    Type currentType = type;
    for (int i=0;i<len;++i) currentType = ((Type.Arr)currentType).result;
    return new NeutralException(currentType, term.apply(rands));
  }
}

File: language/src/main/kotlin/cadenza/nbe/NeutralException.kt
package cadenza.nbe

import cadenza.types.Type
import cadenza.values.NeutralValue
import com.oracle.truffle.api.nodes.SlowPathException

// this is used to suck the program through a straw using normalization-by-evaluation
// to produce a beta-eta-long normal form versions of the resulting program that remains

class NeutralException(val type: Type, val term: Neutral) : SlowPathException() {
  fun get(): NeutralValue {
    return NeutralValue(type, term)
  }

  // TODO: throw NeutralException and return Nothing
  fun apply(rands: Array<Any?>): NeutralException {
    val len = rands.size
    var currentType = type
    for (i in 0 until len) currentType = (currentType as Type.Arr).result
    return NeutralException(currentType, term.apply(rands))
  }

  companion object {
    private val serialVersionUID = 5587798688299594259L
  }
}
